File: langs/c++/truth.cxx
#include <iostream>

using namespace std;

class TruthMachine {

public:
    void enter_true() {
        while (true)
            cout << "1" << endl;
    }

    void enter_false() {
        cout << "0" << endl;
    }
};


int main(void) {

    cout << "\n"
      << "C++ implementation of the Truth Machine\n"
      << "\n"
      << "The Truth Machine is a simple program for learning\n"
      << "a programing language that takes only two inputs.\n"
      << "\n"
      << "* 1: Infinitely outputs \"1\". You can exit using <Ctrl-C>.\n"
      << "* 0: Outputs \"0\" and terminates.\n\n";

    string buf;
    TruthMachine tm;

    bool waiting = true;
    do {
        cout << "Enter 1 or 0: ";
        getline(cin, buf);

        try {
            long val = stol(buf);
            if (val == 1L) {
                tm.enter_true();
                waiting = false;
            }
            else if (val == 0L) {
                tm.enter_false();
                waiting = false;
            }
        }
        catch (exception &e) {
            // discard input and retry
        }
    } while (waiting);

    return 0;
}

File: langs/go/truth.go
package main

import (
	"fmt"
	"strconv"
)

type TruthMachine struct{}

func main() {

	intro := `
Go implementation of the Truth Machine

The Truth Machine is a simple program for learning
a programing language that takes only two inputs.

* 1: Infinitely outputs "1". You can exit using <Ctrl-C>.
* 0: Outputs "0" and terminates.
`
	fmt.Println(intro)

	tm := TruthMachine{}

	for true {
		var input string
		fmt.Print("Enter 1 or 0: ")
		_, err := fmt.Scanf("%s", &input)
		if err != nil {
			continue
		}

		var intputInt int
		intputInt, err = strconv.Atoi(input)
		if err != nil {
			continue
		}

		err = tm.Run(intputInt)
		if err == nil {
			break
		} else {
			continue
		}
	}
}

func (this TruthMachine) Run(value int) error {
	if value == 1 {
		this.ExecTrueAction()
		return nil
	} else if value == 0 {
		this.ExecFalseAction()
		return nil
	} else {
		return fmt.Errorf("invalid input: %d. Only 0 or 1 allowed", value)
	}
}

func (this TruthMachine) ExecTrueAction() {
	for true {
		fmt.Println("1")
	}
}
func (this TruthMachine) ExecFalseAction() {
	fmt.Println("0")
}
